File: leetcode/cpp/combinationSum2.cpp



// leetcode.com  40. Combination Sum II

class Solution
{
  public:
    vector<vector<int>> combinationSum2(vector<int> &candidates, int target)
    {
        if (candidates.empty())
            return {};

        vector<vector<int>> res;
        vector<int> out;

        sort(candidates.begin(), candidates.end());

        helper(candidates, target, 0, out, res);

        return res;
    }

    void helper(vector<int> &candidates, int target, int level,
                vector<int> &out, vector<vector<int>> &res)
    {
        if (target < 0)
            return;

        if (target == 0)
        {
            res.push_back(out);
            return;
        }

        for (int i = level; i < candidates.size(); i++)
        {
            if (i > level && candidates[i] == candidates[i - 1]) continue;

            out.push_back(candidates[i]);

            helper(candidates, target - candidates[i], i + 1, out, res);

            out.pop_back();
        }
    }
};

File: leetcode/java/src/main/java/com/vkeonline/leetcode/p000/CombinationSum2.java


package com.vkeonline.leetcode.p000;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.List;

/**
 * Leetcode [M]: 40. Combination Sum II
 * @author xiaojun.yang
 */
public class CombinationSum2 {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        int len = candidates.length;
        List<List<Integer>> res = new ArrayList<>();
        if (len == 0) {
            return res;
        }

        Arrays.sort(candidates);
        Deque<Integer> path = new ArrayDeque<>(len);
        dfs(candidates, len, 0, target, path, res);
        return res;
    }

    private void dfs(int[] candidates, int len, int begin, int target,
            Deque<Integer> path, List<List<Integer>> res) {
        if (target == 0) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = begin; i < len; i++) {
            if (target - candidates[i] < 0) {
                break;
            }

            if (i > begin && candidates[i] == candidates[i - 1]) {
                continue;
            }

            path.addLast(candidates[i]);
            dfs(candidates, len, i + 1, target - candidates[i], path, res);

            path.removeLast();
        }
    }
}
